src/librustc_mir/build/scope.rs

```diff
@@ -214,20 +215,34 @@ impl<'a,'tcx> Builder<'a,'tcx> {
                       extent: CodeExtent,
                       block: BasicBlock,
                       target: BasicBlock) {
-        let popped_scopes =
-            match self.scopes.iter().rev().position(|scope| scope.extent == extent) {
-                Some(p) => p + 1,
-                None => self.hir.span_bug(span, &format!("extent {:?} does not enclose",
-                                                              extent)),
-            };
-
-        for scope in self.scopes.iter_mut().rev().take(popped_scopes) {
+        let Builder { ref mut scopes, ref mut loop_scopes, ref mut cfg, ref mut hir, .. } = *self;
+
+        let scope_count = 1 + scopes.iter().rev().position(|scope| scope.extent == extent)
+                                                 .unwrap_or_else(||{
+            hir.span_bug(span, &format!("extent {:?} does not enclose", extent))
+        });
+
+        // We merge iterators of regular scopes and loop scopes based on their extents. This code
+        // relies on the assumption that every loop scope will have an extent of some non-popped
+        // scope.
```

Comment:
Another option would be to allow a kind of marker to be pushed into the scopes array, so that as we iterate backwards we'd encounter a "track_break" marker and set the flag to true.